### PR TITLE
GEODE-3126: Adding a query message to the protobuf protocol

### PR DESCRIPTION
**Pull request by upthewaterspout**
Adding an oql query request and response to the new protocol.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/ProtobufSerializationService.java [hunk 1]
```diff
@@ -14,10 +14,13 @@
  */
 package org.apache.geode.internal.protocol.protobuf.v1;
 
+import java.util.List;
+
 import com.google.protobuf.ByteString;
 import com.google.protobuf.NullValue;
 
 import org.apache.geode.annotations.Experimental;
+import org.apache.geode.internal.protocol.protobuf.v1.BasicTypes.EncodedValue;
```
Comment (PivotalSarge) [inline geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/ProtobufSerializationService.java]: Where are `java.util.List` and `BasicTypes.EncodedValue` used in `ProtobufSerializationService`?

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/operations/OqlQueryRequestOperationHandler.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      throws InvalidExecutionContextException, ConnectionStateException, EncodingException,
+      DecodingException {
+    String queryString = request.getQuery();
+    List<EncodedValue> encodedParameters = request.getBindParameterList();
+
+    InternalQueryService queryService = messageExecutionContext.getCache().getQueryService();
+
+    Query query = queryService.newQuery(queryString);
+
+    Object[] bindParameters = new Object[encodedParameters.size()];
```
Comment (PivotalSarge) [inline geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/operations/OqlQueryRequestOperationHandler.java]: For symmetry what about `Object[] decodeBindParameters(List<EncodedValue>)`?

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/registry/ProtobufOperationContextRegistry.java [hunk 3]
```diff
@@ -138,5 +142,10 @@ private void addContexts() {
             // Resource permissions get handled per-function, since they have varying permission
             // requirements.
             new ResourcePermission(ResourcePermission.NULL, ResourcePermission.NULL)));
+    operationContexts.put(MessageTypeCase.OQLQUERYREQUEST,
+        new ProtobufOperationContext<>(ClientProtocol.Message::getOqlQueryRequest,
+            new OqlQueryRequestOperationHandler(),
+            opsResp -> ClientProtocol.Message.newBuilder().setOqlQueryResponse(opsResp),
+            new ResourcePermission(Resource.DATA, Operation.READ)));
```
Comment (PivotalSarge) [inline geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/registry/ProtobufOperationContextRegistry.java]: Even though OQL statements can result in the execution of Java methods, is it standard to use `READ` permissions for queries?

Comment (upthewaterspout) [reply geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/registry/ProtobufOperationContextRegistry.java]: Actually, queries can't invoke arbitrary methods - see GEODE-3247. So READ is the correct permission - that's what the old protocol and gfsh and are checking.

File: geode-protobuf-messages/src/main/proto/v1/region_API.proto [hunk 4]
```diff
@@ -84,3 +84,19 @@ message GetRegionRequest {
...
+    repeated EncodedValue bindParameter = 2;
+}
+
+//Response to an OQL query. Depending on the query,
+//it may return a single value, a list of values, or a table
+//of values.
+message OQLQueryResponse {
+    oneof result {
+      EncodedValue singleResult = 1;
+      List listResult = 2;
```
[the remark this answers is not in the document]
Comment (upthewaterspout) [reply geode-protobuf-messages/src/main/proto/v1/region_API.proto]: Yeah, me too. Unfortunately, protobuf doesn't let you put a repeated inside of a oneof, which is why we need this list message.

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/registry/ProtobufOperationContextRegistry.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (upthewaterspout) [reply geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/registry/ProtobufOperationContextRegistry.java]: Done - GEODE-4750.

**Pull request merged**